File: KisaragiTech.Dape/Base/Strings/StringSplitter.cs
using System;
using System.Collections.Generic;

namespace KisaragiTech.Dape.Base.Strings;

public static class StringSplitter
{
    public static IEnumerable<ReadOnlyMemory<char>> ToMemories(string haystack, char needle)
    {
        var start = 0;
        var index = haystack.IndexOf(needle, start);
        while (index != -1)
        {
            yield return haystack.AsMemory(start, index - start);
            start = index + 1;
            index = haystack.IndexOf(needle, start);
        }

        // 最後のピース
        if (start <= haystack.Length)
            yield return haystack.AsMemory(start);
    }
}

File: src/main/scala/io/github/ksrgtech/dape/base/strings/StringSplitter.scala
package io.github.ksrgtech.dape.base.strings

object StringSplitter {

  def toSubstrings(haystack: String, needle: Char): Iterator[String] =
    new Iterator[String] {
      private var start: Int       = 0
      private var nextIndex: Int   = haystack.indexOf(needle)
      private var hasMore: Boolean = true

      override def hasNext: Boolean = hasMore

      override def next(): String = {
        if !hasMore then {
          throw new NoSuchElementException("No more elements")
        }

        if nextIndex == -1 then {
          hasMore = false
          haystack.substring(start)
        } else {
          val result = haystack.substring(start, nextIndex)
          start = nextIndex + 1
          nextIndex = haystack.indexOf(needle, start)
          result
        }
      }
    }
}
